Temporal: Reimplement TemporalDuration::total() according to the spec

catamorphism opened the pull request:
#### ae596477c51879cd2573d03e828d948a9de500e2
<pre>
Temporal: Reimplement TemporalDuration::total() according to the spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=298561">https://bugs.webkit.org/show_bug.cgi?id=298561</a>

Reviewed by Yusuke Suzuki.

This enables some of the tests for Temporal/Duration/prototype/total.

This patch was co-authored by Philip Chimento &lt;pchimento@igalia.com&gt;.

* JSTests/stress/temporal-duration.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::appendInteger):
(JSC::totalTimeDuration):
(JSC::TemporalDuration::total const):
* Source/JavaScriptCore/runtime/TemporalObject.h:
(JSC::lengthInNanoseconds):

Canonical link: <a href="https://commits.webkit.org/300802@main">https://commits.webkit.org/300802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d2b115f9d6e817e75f61363c4d1c5b4d582dbdf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123920 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130710 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52229 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94255 "") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126874 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/35341 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110852 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74855 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29013 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74183 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116079 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105070 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/29236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133403 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122452 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50873 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102722 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51247 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102549 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47891 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19488 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50726 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56490 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153548 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50200 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39026 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51874 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -559,6 +559,65 @@ ISO8601::Duration TemporalDuration::subtract(JSGlobalObject* globalObject, JSVal
+    // This causes the test262 test
+    // Temporal/Duration/prototype/total/precision-exact-mathematical-values-7.js
+    // to fail when unit=milliseconds, smallerUnit=microseconds, integer=2**53, fraction=1999.
+    return sign * result.toString().toDouble();
```
This looks very inefficient. Why do we need to do this string and re-parsing?

catamorphism replied:
The code is making up for the lack of float128. [The spec](https://tc39.es/proposal-temporal/#sec-temporal-totaltimeduration) calls this out: "The following step cannot be implemented directly using floating-point arithmetic when [𝔽](https://tc39.es/ecma262/#%F0%9D%94%BD)(timeDuration) is not a [safe integer](https://tc39.es/ecma262/#safe-integer)."

I followed [the polyfill implementation](https://github.com/tc39/proposal-temporal/blob/main/polyfill/lib/timeduration.mjs#L92) closely, but if you have a better idea, let me know.

Constellation replied:
Apparently, boa temporal etc. is not doing this inefficient implementation. So, r-.
Translating JS polyfill to C++ is not what we should follow.

catamorphism replied:
Boa fails the test `Temporal/Duration/prototype/total/precision-mathematical-values-6.js`. (see https://boajs.dev/conformance?version=main&testPath=built-ins%2FTemporal%2FDuration%2Fprototype%2Ftotal ). There is a note to this effect [here](https://github.com/boa-dev/temporal/blob/main/src/builtins/core/duration/normalized.rs#L165) ("TODO Fix: Arithemtic on floating point numbers is not safe. According to NOTE 2 in the spec"). 

On the other hand, tests like this are the uncommon case, so I'll try to think about whether there's a way to optimize the common case.

catamorphism replied:
@Constellation I took a look at what SpiderMonkey does; their implementation is conformant (at least for the Duration/prototype/total tests). So I adapted that code instead of using the algorithm from the polyfill. Let me know what you think of this approach.

catamorphism replied:
I've been told it may not be an option to commit non-BSD-licensed code to WebKit, so this approach might not be feasible. In any event, I'll be on vacation until Sept. 29; will reply to any comments when I return.

catamorphism committed:
Temporal: Reimplement TemporalDuration::total() according to the spec
https://bugs.webkit.org/show_bug.cgi?id=298561

Reviewed by Yusuke Suzuki.

This enables some of the tests for Temporal/Duration/prototype/total.

This patch was co-authored by Philip Chimento <pchimento@igalia.com>.

* JSTests/stress/temporal-duration.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::appendInteger):
(JSC::totalTimeDuration):
(JSC::TemporalDuration::total const):
* Source/JavaScriptCore/runtime/TemporalObject.h:
(JSC::lengthInNanoseconds):

Canonical link: https://commits.webkit.org/300802@main
